File: bfs.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def bfs(root):
    if root is None:
        return

    queue = [root]

    while queue:
        node = queue.pop(0)
        print(node.value)

        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

File: dsa/bfs.ts
type TreeNode<T> = {
  val: T;
  left: TreeNode<T> | null;
  right: TreeNode<T> | null;
}

function bfsLevelOrder<T>(root: TreeNode<T> | null): T[] {
  if (!root) return [];

  const result: T[] = [];
  const queue: Array<TreeNode<T>> = [root];

  while (queue) {
    const node = queue.shift();
    if (!node) break;
    
    result.push(node.val);

    if (node.left) queue.push(node.left);
    if (node.right) queue.push(node.right);
  }
  
  return result;  
}

